py: add_array() will not add to kv store if value is an empty array

mofosyne opened the pull request:
Problem in https://github.com/ggerganov/llama.cpp/issues/8769 replicated and traced to add_array() adding an empty array to be written to kv store. We don't allow empty array in kv metadata store.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

mofosyne committed:
gguf_writer.py: add_array() should not add to kv store if empty

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -312,6 +312,8 @@ def add_string(self, key: str, val: str) -> None:
         self.add_key_value(key, val, GGUFValueType.STRING)
 
     def add_array(self, key: str, val: Sequence[Any]) -> None:
+        if not val:
```
```suggestion
        if len(val) == 0:
```

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -845,7 +847,14 @@ def _pack_val(self, val: Any, vtype: GGUFValueType, add_vtype: bool) -> bytes:
+            if not isinstance(val, Sequence):
+                raise ValueError(f"Invalid GGUF metadata array, expecting sequence")
+
+            if not val:
```
```suggestion
            if len(val) == 0:
```

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -845,7 +847,14 @@ def _pack_val(self, val: Any, vtype: GGUFValueType, add_vtype: bool) -> bytes:
+        elif vtype == GGUFValueType.ARRAY:
+
+            if not isinstance(val, Sequence):
+                raise ValueError(f"Invalid GGUF metadata array, expecting sequence")
```
```suggestion
                raise ValueError("Invalid GGUF metadata array, expecting sequence")
```

Otherwise flake8 complains

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -845,7 +847,14 @@ def _pack_val(self, val: Any, vtype: GGUFValueType, add_vtype: bool) -> bytes:
+                raise ValueError(f"Invalid GGUF metadata array, expecting sequence")
+
+            if not val:
+                raise ValueError(f"Invalid GGUF metadata array. Empty array")
```
```suggestion
                raise ValueError("Invalid GGUF metadata array. Empty array")
```

mofosyne committed:
Apply suggestions from code review

I was wondering if there was a specific reason for `if val` but good to hear we can safely use `len(val == 0`

Co-authored-by: compilade <git@compilade.net>

mofosyne commented:
Happy for any maintainers to press merge if all checks passes and they don't spot anything.
